[docs] [#9111] add cloud native dev with Gitpod and Codespaces (#9112)
yugabyte/yugabyte-db: docs/content/latest/develop/gitdev/gitpod.md

---
title: Gitpod
linkTitle: Gitpod
description: Ready to code Gitpod integrated dev environment
menu:
  latest:
    identifier: gitpod
    parent: gitdev
    weight: 592
isTocNested: true
showAsideToc: true
---

Use the [Gitpod](https://www.gitpod.io) workspaces to provision an instant development environment with a pre-configured YugabyteDB.

> Gitpod is a configurable ready to code cloud development environment accessible via a browser. A Gitpod workspace includes everything developers need to develop for a specific repository, including the `vscode` or `theia` editing experience, common languages, tools, and utilities. Instantly it sets up a cloud-hosted, containerized, and customizable vscode/theia environment.

Reviewer: Theia is gone as of April: https://www.theregister.com/2021/04/08/gitpod_talks_up_importance_of/

```suggestion
Gitpod is a configurable ready-to-code cloud development environment accessible via a browser. A Gitpod workspace includes everything you need to develop for a specific repository, including the Visual Studio Code editing experience, common languages, tools, and utilities. Instantly it sets up a cloud-hosted, containerized, and customizable editing environment.
```

> Follow the steps below to set up a Gitpod workspace environment with a pre-configured YugabyteDB. For details on Gitpod workspaces, see the [Gitpod documentation](https://www.gitpod.io/docs/).

Reviewer: ```suggestion
Follow the steps on this page to set up a Gitpod workspace environment with a pre-configured YugabyteDB. For details on Gitpod workspaces, refer to the [Gitpod documentation](https://www.gitpod.io/docs/).
```

description: Gitpod integrated development environment
menu:
  latest:
    identifier: gitpod
    parent: gitdev
    weight: 592
isTocNested: true
showAsideToc: true
---

Use [Gitpod](https://www.gitpod.io) workspaces to provision an instant development environment with a pre-configured YugabyteDB.

Gitpod is a configurable ready-to-code cloud development environment accessible via a browser. A Gitpod workspace includes everything you need to develop for a specific repository, including the Visual Studio Code editing experience, common languages, tools, and utilities. Instantly this sets up a cloud-hosted, containerized, and customizable editing environment.

Follow the steps on this page to set up a Gitpod workspace environment with a pre-configured YugabyteDB. For details on Gitpod workspaces, refer to the [Gitpod documentation](https://www.gitpod.io/docs/).

## Requirements

Gitpod doesn't require anything in your local workstation other than a code editor and Git CLI. Much of the development happens in the cloud through a web browser.

## Get started with a boot app

You can find the source at [Spring Boot todo on GitHub](https://github.com/yugabyte/yb-todo-app.git). 

> The easy way to get started with Gitpod is to simply fork this [source repo](https://github.com/yugabyte/yb-todo-app.git) and initialize the Gitpod workspace environment by invoking `https://gitpod.io/#[REPO_URL]` in a browser window. Replace `[REPO_URL]` with your forked repository URL and you should already have been conneted to the Gitpod account before launching the browser URL.

Reviewer: "...should already have been connected..." - does this mean I should be logged into my Gitpod account before I do this step?

Reviewer: Also, if i'm doing the easy way, do i need to follow any of the instructions on this page other than "fork the repo, browse to gitpod.io/#repo_url"? I suspect we need to say "do this, then skip to [Customize the Gitpod environment]"

Author: > "...should already have been connected..." - does this mean I should be logged into my Gitpod account before I do this step?
Yes, u need to be logged in. If not, you would be taken through the login flow.

> Also, if i'm doing the easy way, do i need to follow any of the instructions on this page other than "fork the repo, browse to gitpod.io/#repo_url"? I suspect we need to say "do this, then skip to [Customize the Gitpod environment]"
Easy way => just fork and launch the URL

> If you want **to set this up from scratch**, use the following instructions to bootstrap the base project template and copy the appropriate files/content from the [source repo](https://github.com/yugabyte/yb-todo-app.git).

Reviewer 2: ```suggestion
If you want **to set the Spring Boot app from scratch**, use the following instructions to bootstrap the base project template and copy the appropriate files/content from the [source repo](https://github.com/yugabyte/yb-todo-app.git).
```

### Initialize the base project structure

Spring todo is a Java Spring Boot reactive app. However, the steps to go through the Gitpod experience are agnostic of the language/framework. A quick way to get started with a spring boot app is via the [Spring Initializer](https://start.spring.io). Generate the base project structure with Webflux, Flyway, and R2DBC dependencies.

![Set up the base project abstract](/images/develop/gitdev/gitpod/init-sb.png)

### Complete the CRUD APIs

> Complete the todo-service by copying the source and build files from the [source repo](https://github.com/yugabyte/yb-todo-app.git) to your repo to handle GET, POST, PUT, and DELETE API requests.

Reviewer 2: ```suggestion
Complete the todo-service by copying the source and build files from the [source repo](https://github.com/yugabyte/yb-todo-app.git) to your repo to handle GET, POST, PUT, and DELETE API requests. Once you have your project ready, make sure to create a repository on Github and add your project to this repository.
```

![Complete the API endpoints](/images/develop/gitdev/gitpod/complete-api.png)

{{< note title="Note" >}}
The application uses non-blocking reactive APIs to connect to YugabyteDB.
{{< /note >}}

## Initialize Gitpod

To get started quickly, you can use the universal image [pre-built containers](https://www.gitpod.io/docs/quickstart) or a language-specific image. It can be further customized to fit your needs either by extending them or by creating a new one. A simple click provisions the entire development environment in the cloud with an integrated powerful Visual Studio Code editor. The entire config to set up the development environment lives in the same source code repository. Follow the steps in the next sections to set up your Gitpod environment.

### Set up the Gitpod environment

You initialize the Gitpod workspace environment for a specific repository by invoking `https://gitpod.io/#[REPO_URL]`.

![Initalize the workspace environment](/images/develop/gitdev/gitpod/init-workspace.png)

> You can either use a universal image with pre-configured libraries and commonly used utilities or a language-specific image. Create the integrated YugabyteDB workspace environment by customizing the base universal image.

Reviewer 2: ```suggestion
You can either use a universal image with pre-configured libraries and commonly used utilities or a language-specific image. Create the integrated YugabyteDB workspace environment by customizing the base universal image. Use the editor in your gitpod environment to add these files directly and commit them back to your github repository.
```

```yml
# Refer the customized docker image
image:
  file: .gitpodcontainer/Dockerfile

# Run the app and db related jobs
# - run the db post-init script
# - start the DB instance
# - run the db app specific init script
# - build and run the app process
tasks:
  - name: db-run
    before: /usr/local/yugabyte/bin/post_install.sh
    init: |
      yugabyted start --base_dir=$STORE/ybd1 --listen=$LISTEN && \
      [[ ! -f $STORE/.init-db.sql.completed ]] &&  { for i in {1..10}; do (nc -vz $LISTEN $PORT >/dev/null 2>&1); [[ $? -eq 0 ]] &&  { ysqlsh -f $STORE/init-db.sql; touch $STORE/.init-db.sql.completed; break; } || sleep $i; done } && \
      [[ ! -f $STORE/.init-db.sql.completed ]] && echo 'YugabyteDB is not running!'
  - name: app-run
    init: gradle clean build -x test
    command: java -jar build/libs/*.jar

ports:
  - port: 8080
    onOpen: notify
  - port: 7000
    onOpen: notify
  - port: 9000
    onOpen: notify
  - port: 37843
    onOpen: ignore
  - port: 7100
    onOpen: ignore
  - port: 9100
    onOpen: ignore
  - port: 5433
    onOpen: ignore
  - port: 13000
    onOpen: ignore
  - port: 9042
    onOpen: ignore
  - port: 12000
    onOpen: ignore    
```

> Launch the workspaces environment with this updated spec to provision the development environment with a running YugabyteDB instance. It opens two terminals, one terminal to run the DB task, and the other to compile and run the boot app.

Reviewer 2: ```suggestion
Make sure to commit the above files to your Github repository from your existing gitpod environment. 
Next, [launch](../../../latest/develop/gitdev/gitpod/#set-up-the-gitpod-environment) the workspaces environment again with this updated spec to provision the development environment with a running YugabyteDB instance. It opens two terminals, one terminal to run the DB task, and the other to compile and run the boot app.
```
